andela/codepirates-ah-backend
# #167313417 Users should be able to bookmark articles for later reading

#### What does this PR do?

  Add bookmarks feature

#### Description of Task to be completed?

  - Implement basic create read update and delete bookmarks functionalities
  - Grouping of bookmarks in collections
  - Duplication of bookmarks

#### How should this be manually tested?

  With a base url of https://codepirates-ah-backend.herokuapp.com/api/v1/users for remote and http://localhost:port/api/v1/users for local testing, send the following requests the *requests are authenticated*:

  - **Create a bookmark** POST/bookmarks with a body a son object having properties of name and articleId. The articleId must be an integer, the name could be empty or any string
    *Possible outcomes:*
    - With a valid id of an existing article, a bookmark should be created with the supplied name if not empty or the article title appended with a timestamp if current article not bookmarked by current user before
    - Try creating the same bookmark again, changing alternately the name and articleId values, randomly or targeted and examine the possible outcomes

  - **Update bookmark** PATCH/bookmarks/*target-bookmark-name* supplying a json object having the desired new name as property name and the articleId. Try all possible combinations of these three parameters, random, repeating and specific to explore the different behaviours of the application

  - **Get a specific bookmark** GET/bookmarks/*target-bookmark-name* 

  - **Get all bookmarks of the current user** GET/bookmarks

  - **Delete a specific bookmark** DELETE/bookmarks/*target-bookmark-name*

  - **Delete all bookmarks of the current user** DELETE/bookmarks

  - **Add a bookmark to a collection** POST/bookmarks/collections with a body having the properties collection for collection name, name for bookmark name and articleId

  - **Change a collection's name** PATCH/bookmarks/collections/*target-collection-name* supplying the desired new name as value to a collection property in the request body

  - **View a specific collection** GET/bookmarks/collections/*target-collection-name*

  - **Get all of user's collections** GET/bookmarks/collections

  - **Delete a bookmark from a collection** DELETE/bookmarks/collections/*target-collection*/*target-bookmark*

  - **Delete a whole collection** DELETE/bookmarks/collections/*collection-name*

#### Any background context you want to provide?

#### What are the relevant pivotal tracker stories?

  [finishes #167313417](https://www.pivotaltracker.com/story/show/167313417)
#### Screenshots (if appropriate)

Create bookmark
<img width="1123" alt="Screen Shot 2019-08-27 at 22 50 38" src="https://user-images.githubusercontent.com/36167006/63807027-34e26e80-c91d-11e9-8ff6-2a2232bd65ff.png">

Change a bookmark's name
<img width="1116" alt="Screen Shot 2019-08-27 at 22 52 33" src="https://user-images.githubusercontent.com/36167006/63807159-7bd06400-c91d-11e9-9dd3-e9d53de8f272.png">

Get a specific bookmark
<img width="1122" alt="Screen Shot 2019-08-27 at 22 54 22" src="https://user-images.githubusercontent.com/36167006/63807316-bcc87880-c91d-11e9-835a-6f5d5ef03a27.png">


Delete all bookmarks
<img width="1125" alt="Screen Shot 2019-08-27 at 22 56 08" src="https://user-images.githubusercontent.com/36167006/63807417-f305f800-c91d-11e9-98bb-f282370a7591.png">


#### Questions:
